### ISSUE TITLE
read ECONNRESET error using aws-iot-device-sdk

### ISSUE THREAD

**Issue by gauthamkantharaju** [aws/aws-iot-device-sdk-js#253]
Hi,

Using aws-iot-device-sdk npm package (version 2.2.1) for publishing data to cloud. Application (mqtt client) code publishes data to cloud every 15 minutes interval.

```
device = awsIot.device({
			keyPath: process.env.AWS_MQTT_KEY,
			certPath: process.env.AWS_MQTT_CERT,
			caPath: process.env.AWS_MQTT_ROOTCA,
			clientId: process.env.AWS_MQTT_CLIENT_ID,
			region: process.env.AWS_MQTT_REGION,
			host: process.env.AWS_MQTT_HOST,
			offlineQueueMaxSize: parseInt(process.env.AWS_MQTT_OFFLINE_QUEUE_MAX_SIZE),
			offlineQueueDropBehavior: process.env.AWS_MQTT_OFFLINE_QUEUE_DROP_BEHAVIOR,
			maximumReconnectTimeMs: parseInt(process.env.AWS_MQTT_MAX_RECONNECT_TIME_MSEC),
			keepalive: parseInt(process.env.AWS_MQTT_KEEPALIVE_SEC)
		});

```

Using device.on("error"),

device.on("error", function (err) {
			console.log("DNS issue or AWS IoT sdk error: " + err);
});

seeing below error very often from last week,

DNS issue or AWS IoT sdk error: **Error: read ECONNRESET** and connects to AWS IoT. This error is seen irrespective of keepalive value i.e. when keepalive is '0' or '>0'

Is this a known issue? else let me know why I am seeing this error?

Thank you in advance,
Gautham

**Comment by justinboswell** [comment]
Have you previously been able to connect successfully and maintain a connection?

**Comment by gauthamkantharaju** [comment]
@justinboswell yes previously was able to connect successfully and had a stable connection for months.

**Comment by justinboswell** [comment]
Ok, good to know. Have you made any other recent changes to your policy or account that might account for the change? Have you pulled new npm dependencies or made any other environment changes?

**Comment by gauthamkantharaju** [comment]
No changes in your policy or account.  Did install same node and npm binaries which I used previously and installed the npm packages (fixed versions) as in the package.json.

There was no aws-iot-device-sdk npm package version change right? was there any change in aws-iot-device-sdk npm dependencies?

No environmental changes.

**Comment by justinboswell** [comment]
The js SDK hasn't changed in a year. If you can't resolve it locally, I would suggest contacting AWS support through the console.

**Comment by gauthamkantharaju** [comment]
> The js SDK hasn't changed in a year. If you can't resolve it locally, I would suggest contacting AWS support through the console.

Thank you for the quick response. Sure will contact AWS support.

**Comment by AnkushBhalage** [comment]
Facing same problem?

**Comment by AnkushBhalage** [comment]
please provide solution

**Comment by gauthamkantharaju** [comment]
@AnkushBhalage ECONNRESET happens if the network connection is not stable and make sure keepalive is min 300 secs, which I guess is the default value, not sure!

**Comment by AnkushBhalage** [comment]
Thanks @gauthamkantharaju Issue has been solved.

Code:
var AWSIot= require('aws-iot-device-sdk')

var device =AWSIot.device({
    keyPath: 'device-private.pem.key',
    certPath: 'device.pem.crt',
      caPath: 'RootCA.pem',
    clientId: 'FEV_1',
        host: 'a36o9z4m44q7dt-ats.iot.us-east-1.amazonaws.com',
        keepalive:300
})

device.on("error", function (err) {
    console.log("Error: " + err);
    });
  device
    .on('connect', function() {
      console.log('connect');
       device.subscribe('peter');
      // device.publish('peter', JSON.stringify({ test_data: "Ankush"}));
      
    });
  
    device
  .on('message', function(topic, payload) {
    console.log('message', topic, payload.toString());
  });

**Comment by gauthamkantharaju** [comment]
@AnkushBhalage you're welcome and good to hear that your issue is resolved :)
